### PR TITLE
Cleaning up tests and round constants

### PR DESCRIPTION
**Pull request by SyxtonPrime**
Removed the import of HorizenLabs Poseidon2 implementation from BabyBear and Goldilocks tests. Instead saved the relevent constants in a file in Poseidon2 and computed a couple of test evaluations using sage to check results again.

I haven't modified the bn254 tests as that version of Poseidon2 is a little different (Width is not a multiple of 4) so I don't currently have a sage implementation of it. Not sure how much we are planning to support bn254 anyway so just going to leave that as is for now.

Additionally, minor changes to the Poseidon2 algorithm where I have split apart the constants into external and internal round constants. The external constants are each the form [F; WIDTH] where as the internal ones are directly elements of F so it doesn't really make sense to save the constants together.

### CODE REVIEW

File: poseidon2/src/round_constants.rs [hunk 1]
```diff
@@ -0,0 +1,174 @@
+// We save the constants used by the Poseidon2 implementation by Horizon Labs:
```
Comment (jorgeantonio21) [inline poseidon2/src/round_constants.rs:1]: nit: Horizen*

**Review by jorgeantonio21 — COMMENTED** [review]
LGTM

File: poseidon2/src/round_constants.rs (hunk 1 above)
Comment (SyxtonPrime) [reply poseidon2/src/round_constants.rs]: Good catch, just fixed that.

File: bn254-fr/src/poseidon2.rs [hunk 2]
```diff
@@ -89,9 +89,32 @@ mod tests {
...
+                }
+            })
+            .collect();
+        let internal_round_constants: Vec<F> = round_constants
+            .iter()
+            .skip(ROUNDS_F / 2)
+            .take(ROUNDS_P)
+            .map(|vec| vec[0])
+            .collect();
+
```
Comment (unzvfu) [inline bn254-fr/src/poseidon2.rs:109]: Shorter, clearer, and more idiomatic:
```suggestion
        let internal_start = ROUNDS_F / 2;
        let internal_end = (ROUNDS_F / 2) + ROUNDS_P;
        // need to make `round_constants` mutable above for this to work
        let internal_round_constants = round_constants
            .drain(internal_start..internal_end)
            .map(|vec| vec[0])
            .collect::<Vec<_>>();
        // optional rename; probably simpler to just to call the original variable `external_round_constants`
        let external_round_constants = round_constants;
```

**Review by unzvfu — CHANGES_REQUESTED** [review]
Looks good overall, just a few minor things that can be improved.

File: poseidon2/src/round_constants.rs [hunk 3]
```diff
@@ -0,0 +1,174 @@
+// We save the constants used by the Poseidon2 implementation by Horizen Labs:
+// https://github.com/HorizenLabs/poseidon2
+// This lets us test that our implementations match theirs.
+
+// Note that for the external (full) rounds, their implementation uses the matrix:
+// [[5, 7, 1, 3], [4, 6, 1, 1], [1, 3, 5, 7], [1, 1, 4, 6]]
```
[the remark this answers is not in the document]
Comment (SyxtonPrime) [reply poseidon2/src/round_constants.rs:6]: In Poseidon2, the external (full) round matrix is built from a 4x4 MDS matrix. Happy to reword this comment though.

File: poseidon2/src/lib.rs [hunk 4]
```diff
@@ -67,16 +74,21 @@ where
...
+            internal_constants.push(rng.gen::<F>());
         }
 
         Self {
             rounds_f,
+            external_constants,
             rounds_p,
-            constants,
+            internal_constants,
             internal_linear_layer: internal_mds,
```
[the remark this answers is not in the document]
Comment (SyxtonPrime) [reply poseidon2/src/lib.rs:92]: I completely agree. I might make this change in a future PR though as I'll introduce external_linear_layer as an argument and it's natural to fix this at the same time.

File: poseidon2/src/lib.rs [hunk 5]
```diff
@@ -67,16 +74,21 @@ where
...
-        for _ in 0..rounds {
-            constants.push(rng.gen::<[F; WIDTH]>());
+        let mut external_constants = Vec::new();
+        for _ in 0..rounds_f {
+            external_constants.push(rng.gen::<[F; WIDTH]>());
+        }
+
+        let mut internal_constants = Vec::new();
+        for _ in 0..rounds_p {
+            internal_constants.push(rng.gen::<F>());
```
[the remark this answers is not in the document]
Comment (SyxtonPrime) [reply poseidon2/src/lib.rs:84]: Sounds good

File: bn254-fr/src/poseidon2.rs [hunk 6]
```diff
@@ -89,24 +89,15 @@ mod tests {
...
-                }
-            })
-            .collect();
-        let internal_round_constants: Vec<F> = round_constants
-            .iter()
-            .skip(ROUNDS_F / 2)
-            .take(ROUNDS_P)
+        let internal_start = ROUNDS_F / 2;
+        let internal_end = (ROUNDS_F / 2) + ROUNDS_P;
+        // need to make `round_constants` mutable above for this to work
```
Comment (unzvfu) [inline bn254-fr/src/poseidon2.rs:94]: Don't need this comment anymore.

File: bn254-fr/src/poseidon2.rs [hunk 7]
```diff
@@ -89,24 +89,15 @@ mod tests {
...
+        let internal_round_constants = round_constants
+            .drain(internal_start..internal_end)
             .map(|vec| vec[0])
-            .collect();
-
+            .collect::<Vec<_>>();
+        // optional rename; probably simpler to just to call the original variable `external_round_constants`
```
Comment (unzvfu) [inline bn254-fr/src/poseidon2.rs:99]: Don't need this comment anymore.

**Review by unzvfu — APPROVED** [review]
Looks good! Just a few leftover comments to remove.

**Pull request merged**